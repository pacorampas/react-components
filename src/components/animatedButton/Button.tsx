import React, { useState } from 'react'
import styled, { css } from 'styled-components'
import { focusBackgroundPseudoStates } from '../../styledHelpers'
import { ThemeProp, VARIANT } from '../../theme/theme.types'
import { IncreaseAnim, ANIM_STATUS } from '../increaseAnim/IncreaseAnim'

export interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  variant?: VARIANT
  /**
   * Use a button without background
   */
  bordered?: boolean
  /**
   * Button contents
   */
  children: JSX.Element | string
  /**
   * Optional click handler
   */
  onClick?: () => void
}

const defaultProps = {
  variant: VARIANT.PRIMARY,
  bordered: false,
}

const Wrapper = styled.button`
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;

  ${({ theme, variant = defaultProps.variant, bordered = defaultProps.bordered }: ButtonProps & ThemeProp) => css`
    ${focusBackgroundPseudoStates({ theme, variant, bordered })}
    ${theme?.components?.button?.overrides}
  `};
`

const Label = styled.span`
  position: relative;
  font-family: 'Poppins';
`

/**
 * Primary UI component for user interaction
 */
export const Button = ({
  variant = defaultProps.variant,
  children,
  onClick,
  ...rest
}: ButtonProps): React.ReactElement => {
  const [anim, setAnim] = useState(false)

  const handleClick = () => {
    setAnim(true)
    onClick?.()
  }

  const handleIncreaseAnimStop = () => {
    setAnim(false)
  }

  return (
    <Wrapper {...{ variant }} onClick={handleClick} {...rest}>
      <>
        {anim && (
          <IncreaseAnim
            variant={variant as VARIANT}
            status={ANIM_STATUS.PLAY}
            increase={20}
            onStop={handleIncreaseAnimStop}
          />
        )}
        <Label>{children}</Label>
      </>
    </Wrapper>
  )
}
