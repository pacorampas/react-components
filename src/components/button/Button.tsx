import React from 'react'
import styled, { css } from 'styled-components'
import { ThemeProp, VARIANT } from '../../theme'

export interface ButtonProps extends ThemeProp, React.HTMLAttributes<HTMLButtonElement> {
  className?: string
  /**
   * The id of the `button` element.
   */
  id?: string
  /**
   * Is button disabled?
   */
  disabled?: boolean
  /**
   * Button children
   */
  children: React.ReactChild | string
  /**
   * Button variant
   */
  variant: VARIANT
  /**
   * Is button bordered?
   */
  bordered?: boolean
}

const StyledButton = styled.button<Partial<ButtonProps>>`
  position: relative;
  font-family: 'Poppins';
  font-size: ${({ theme }) => theme.components.button.fontSize};
  background-color: ${({ theme, variant, bordered }) => {
    const color = theme.colors[variant ?? VARIANT.PRIMARY]
    return bordered ? 'transparent' : color.main
  }};
  color: ${({ theme, variant, bordered }) => {
    const color = theme.colors[variant ?? VARIANT.PRIMARY]
    return bordered ? color.main : color.contrast
  }};
  height: ${({ theme }) => theme.components.button.height};
  padding: ${({ theme }) => theme.components.button.padding};
  transition: all 0.3s ease;
  cursor: pointer;
  user-select: none;

  ${({ theme, variant }) => {
    const { style, width, radius } = theme?.components?.button?.border ?? {}

    return css`
      border-style: ${style};
      border-width: ${width};
      border-color: ${theme.colors[variant ?? VARIANT.PRIMARY]?.main};
      border-radius: ${radius};

      &:not([disabled]):hover,
      &:not([disabled]):focus,
      &:not([disabled]):active {
        box-shadow: ${() => `0 0 0 1px ${theme.colors[variant ?? VARIANT.PRIMARY]?.main}`};
        opacity: ${({ theme, bordered }: Partial<ButtonProps>) =>
          !bordered && `${theme?.components?.button?.activeOpacity}`};
        background-color: ${({ theme, variant, bordered }: Partial<ButtonProps>) =>
          bordered && theme?.colors[variant ?? VARIANT.PRIMARY].main};
        color: ${({ theme, variant, bordered }: Partial<ButtonProps>) =>
          bordered && theme?.colors[variant ?? VARIANT.PRIMARY].contrast};
      }
    `
  }};

  &:disabled {
    cursor: auto;
    opacity: ${({ theme }) => theme.components.button.disabledOpacity};
  }
`

export const Button = ({ variant, bordered, children, ...rest }: ButtonProps): React.ReactElement => (
  <StyledButton variant={variant} bordered={bordered} {...rest}>
    {children}
  </StyledButton>
)

Button.defaultProps = {
  type: 'button',
  disabled: false,
  bordered: false,
}
