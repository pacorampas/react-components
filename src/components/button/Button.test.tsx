import React from 'react'
import { render } from '@testing-library/react'

import { Button } from './Button'

describe('<Button />', () => {
  test('should print label text', async () => {
    const { getByRole } = render(<Button>Call to action</Button>)

    const button = getByRole('button')

    expect(button).toContainHTML('Call to action')
  })
})
