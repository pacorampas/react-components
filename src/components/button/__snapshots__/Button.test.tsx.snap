// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button with modifiers should render a bordered button 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: transparent;
  color: pink;
  border-color: pink;
}

.c0:hover {
  background: transparent;
  border-color: #ffd4df;
  color: #ffd4df;
}

.c0:active {
  background: transparent;
  border-color: #ffdee9;
  color: #ffdee9;
}

.c0:focus {
  border-color: #e5a7b2;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render a solid button 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: pink;
  color: black;
  border-color: pink;
}

.c0:hover {
  background: #ffd4df;
  border-color: #ffd4df;
  color: black;
}

.c0:active {
  background: #ffdee9;
  border-color: #ffdee9;
  color: black;
}

.c0:focus {
  border-color: #e5a7b2;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render using info 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: blue;
  color: white;
  border-color: blue;
}

.c0:hover {
  background: #4221ff;
  border-color: #4221ff;
  color: white;
}

.c0:active {
  background: #542dff;
  border-color: #542dff;
  color: white;
}

.c0:focus {
  border-color: #0000e4;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render using negative 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: red;
  color: white;
  border-color: red;
}

.c0:hover {
  background: #ff3219;
  border-color: #ff3219;
  color: white;
}

.c0:active {
  background: #ff4123;
  border-color: #ff4123;
  color: white;
}

.c0:focus {
  border-color: #e00000;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render using positive 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: green;
  color: white;
  border-color: green;
}

.c0:hover {
  background: #28931c;
  border-color: #28931c;
  color: white;
}

.c0:active {
  background: #359d27;
  border-color: #359d27;
  color: white;
}

.c0:focus {
  border-color: #006900;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render using primary 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: pink;
  color: black;
  border-color: pink;
}

.c0:hover {
  background: #ffd4df;
  border-color: #ffd4df;
  color: black;
}

.c0:active {
  background: #ffdee9;
  border-color: #ffdee9;
  color: black;
}

.c0:focus {
  border-color: #e5a7b2;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;

exports[`Button with modifiers should render using warning 1`] = `
.c0 {
  position: relative;
  outline: 0;
  border: solid 2px transparent;
  border-radius: 4px;
  padding: 8px 12px;
  cursor: pointer;
  background: orange;
  color: white;
  border-color: orange;
}

.c0:hover {
  background: #ffb827;
  border-color: #ffb827;
  color: white;
}

.c0:active {
  background: #ffc233;
  border-color: #ffc233;
  color: white;
}

.c0:focus {
  border-color: #e28d00;
}

.c1 {
  position: relative;
}

<button
  class="c0"
>
  <span
    class="c1"
  >
    Call to action
  </span>
</button>
`;
